client-websocket/src/main/java/com/github/twitch4j/client/websocket/WebsocketConnectionConfig.java

```diff
@@ -0,0 +1,101 @@
+import com.github.twitch4j.common.util.ThreadUtils;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.experimental.Accessors;
+import org.apache.commons.lang3.RandomStringUtils;
+
+import java.time.Duration;
+import java.util.Objects;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Consumer;
+
+@Accessors(chain = true, fluent = true, prefix = "")
+@Setter
+@Getter
+public class WebsocketConnectionConfig {
+
+    public static WebsocketConnectionConfig process(Consumer<WebsocketConnectionConfig> spec) {
+        WebsocketConnectionConfig data = new WebsocketConnectionConfig();
+        spec.accept(data);
+        data.validate();
+        return data;
+    }
+
+    /**
+     * validate the config
+     */
+    public void validate() {
+        Objects.requireNonNull(baseUrl, "baseUrl is a required parameter!");
+        Objects.requireNonNull(wsPingPeriod, "wsPingPeriod is a required parameter, set to 0 to disable!");
+        Objects.requireNonNull(taskExecutor, "taskExecutor is a required parameter!");
+        Objects.requireNonNull(backoffStrategy, "backoffStrategy is a required parameter!");
+    }
+
+    /**
+     * The websocket url for the chat client to connect to.
+     */
+    private String baseUrl;
+
+    /**
+     * WebSocket RFC Ping Period in ms (0 = disabled)
+     */
+    private int wsPingPeriod = 0;
+
+    private ScheduledExecutorService taskExecutor = ThreadUtils.getDefaultScheduledThreadPoolExecutor("twitch4j-" + RandomStringUtils.random(4, true, true), 2);
+
+    /**
+     * Helper class to compute delays between connection retries
+     */
+    private ExponentialBackoffStrategy backoffStrategy = ExponentialBackoffStrategy.builder()
+        .immediateFirst(true)
+        .baseMillis(Duration.ofSeconds(1).toMillis())
+        .jitter(true)
+        .multiplier(2.0)
+        .maximumBackoff(Duration.ofMinutes(5).toMillis())
+        .build();
+
+    /**
+     * called before connecting
+     */
+    private Runnable onPreConnect;
+
+    /**
+     * called after connecting
+     */
+    private Runnable onPostConnect;
+
+    /**
+     * called after the connection to the websocket server has been established successfully
+     */
+    private Runnable onConnected;
+
+    /**
+     * called when receiving a text message on from the websocket server
+     */
+    private Consumer<String> onTextMessage;
+
+    /**
+     * called when connection state is changing from CONNECTED to DISCONNECTING
+     */
+    private Runnable onDisconnecting;
+
+    /**
+     * called before the disconnect
+     * <p>
+     * this occurs after onDisconnecting and before the connection is disposed
+     */
+    private Runnable onPreDisconnect;
+
+    /**
+     * called after the disconnect
+     */
+    private Runnable onPostDisconnect;
+
+    private String proxyHost;
+    private Integer proxyPort;
+    private String proxyUsername;
+    private String proxyPassword;
```

Comment:
I tried to make `client-websocket` fully standalone. Right now the only reason it still depends on `twitch4j-common` is that i wasn't sure what do do about the `backoffStrategy` yet - but if we keep the dep on common we can also use the `ProxyConfig`.

Reply:
To go standalone, we could:

* Create `IBackoffStrategy` interface in the websocket module (would have methods `get()`, `reset()`, `sleep()`)
* Remove explicit references of `ExponentialBackoffStrategy` in `WebsocketConnection`/`WebsocketConnectionConfig` (note: this will require people to pass a concrete implementation when making the config so the backoff strategy is not null)
* Create an adapter class that implements `IBackoffStrategy` & takes `ExponentialBackoffStrategy` in the constructor


As a result, usage would look like:
```kotlin
WebsocketConnection() {
	it.baseUrl("https://twitch4j.github.io")
	it.backoffStrategy(ExponentialBackoffAdapter(ExponentialBackoffStrategy.builder().build()))
}
```


Up to you if this route is desirable

Reply:
I'd still like to make this module useable standalone, even outside of t4j.

I was thinking along the lines of moving some of the useful classes (`ProxyConfig`, `ExponentialBackoffStrategy`, ...) into a `twitch4j-utils` module that has no dependencies itself. I just don't want to use the current common because it includes dependencies (auth, events4j, jackson and other stuff) that shouldn't end up in the websocket client.

Reply:
utils module + interface sounds good!

Shall we directly depend on the classes for now & move them on the next major version bump?
Or could we get away with making common depend on utils module (right now) & move the classes such that the package names are the same (so we virtually avoid the breaking change for most folks)